python/semantic_kernel/functions/kernel_plugin.py

```diff
@@ -28,102 +57,489 @@ class KernelPlugin(KernelBaseModel):
+            for _, candidate in candidates
+            if hasattr(candidate, "__kernel_function__")
+        ]
+        return cls(name=plugin_name, description=description, functions=functions)  # type: ignore
+
+    @classmethod
+    def from_directory(
+        cls,
+        plugin_name: str,
+        parent_directory: str,
+        description: str | None = None,
+    ) -> "KernelPlugin":
+        """Create a plugin from a specified directory.
 
-    def __getitem__(self, name: str) -> "KernelFunction":
-        """Define the [] operator for the plugin
+        This method does not recurse into subdirectories beyond one level deep from the specified plugin directory.
+        For YAML files, function names are extracted from the content of the YAML files themselves (the name property).
+        For directories, the function name is assumed to be the name of the directory. Each KernelFunction object is
+        initialized with data parsed from the associated files and added to a list of functions that are then assigned
+        to the created KernelPlugin object.
+        A native_function.py file is parsed and imported as a plugin,
+        other functions found are then added to this plugin.
+
+        Example:
+            Assuming a plugin directory structure as follows:
+        MyPlugins/
+            |--- pluginA.yaml
+            |--- pluginB.yaml
+            |--- native_function.py
+            |--- Directory1/
+                |--- skprompt.txt
+                |--- config.json
+            |--- Directory2/
+                |--- skprompt.txt
+                |--- config.json
+
+            Calling `KernelPlugin.from_directory("MyPlugins", "/path/to")` will create a KernelPlugin object named
+                "MyPlugins", containing KernelFunction objects for `pluginA.yaml`, `pluginB.yaml`,
+                `Directory1`, and `Directory2`, each initialized with their respective configurations.
+                And functions for anything within native_function.py.
 
         Args:
-            name (str): The name of the function to retrieve.
+            plugin_name (str): The name of the plugin, this is the name of the directory within the parent directory
+            parent_directory (str): The parent directory path where the plugin directory resides
+            description (str | None): The description of the plugin
 
         Returns:
-            The function if it exists, None otherwise.
+            KernelPlugin: The created plugin of type KernelPlugin.
 
         Raises:
-            KeyError: If the function does not exist.
+            PluginInitializationError: If the plugin directory does not exist.
+            PluginInvalidNameError: If the plugin name is invalid.
         """
-        if name not in self.functions:
-            raise KeyError(f"Function {name} not found.")
-        return self.functions[name]
+        MODULE_NAME = "native_function"
```

Comment:
Can we update this now? Why can't we accept any module name and grab the kernel functions from it, if any,